flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -60,89 +63,73 @@
                         taskName,
                         new ChannelStateWriteRequestDispatcherImpl(
                                 taskName,
                                 subtaskIndex,
                                 streamFactoryResolver,
-                                new ChannelStateSerializerImpl())),
-                maxCheckpoints);
+                                new ChannelStateSerializerImpl())));
     }
 
-    ChannelStateWriterImpl(
-            String taskName,
-            ConcurrentMap<Long, ChannelStateWriteResult> results,
-            ChannelStateWriteRequestExecutor executor,
-            int maxCheckpoints) {
+    ChannelStateWriterImpl(String taskName, ChannelStateWriteRequestExecutor executor) {
         this.taskName = taskName;
-        this.results = results;
-        this.maxCheckpoints = maxCheckpoints;
         this.executor = executor;
+        this.ongoingCheckpointId = 0;
+        this.maxAbortedCheckpointId = 0;
     }
 
     @Override
     public void start(long checkpointId, CheckpointOptions checkpointOptions) {
         LOG.debug("{} starting checkpoint {} ({})", taskName, checkpointId, checkpointOptions);
-        ChannelStateWriteResult result = new ChannelStateWriteResult();
-        ChannelStateWriteResult put =
-                results.computeIfAbsent(
+        synchronized (lock) {
+            Preconditions.checkState(checkpointId > ongoingCheckpointId);
+            if (isCheckpointSubsumedOrAborted(checkpointId)) {
+                LOG.debug(
+                        "The checkpoint {} of task {} has been aborted, so don't start.",
                         checkpointId,
-                        id -> {
-                            Preconditions.checkState(
-                                    results.size() < maxCheckpoints,
-                                    String.format(
-                                            "%s can't start %d, results.size() > maxCheckpoints: %d > %d",
-                                            taskName,
-                                            checkpointId,
-                                            results.size(),
-                                            maxCheckpoints));
-                            enqueue(
-                                    new CheckpointStartRequest(
-                                            checkpointId,
-                                            result,
-                                            checkpointOptions.getTargetLocation()),
-                                    false);
-                            return result;
-                        });
-        Preconditions.checkArgument(
-                put == result,
-                taskName + " result future already present for checkpoint " + checkpointId);
+                        taskName);
+                return;
+            }
+            if (result != null) {
+                result.fail(new CancellationException("Cancel old checkpoint."));
```

Comment:
I think it would be safer to fail the futures outside of this lock. We don't know what other locks will be acquired from the actions that are connected to those futures.

So something like:

```
@Nullable ChannelStatWriteResult toFail = null;

synchronized (lock) { 
  ...
  toFail = result;
  ...
}

if (toFail != null) {
  toFail.fail();
}
```
?

Or maybe for the sake of being consistent use the same pattern as in other places?

```
 enqueue(ChannelStateWriteRequest.abort(...));
```

as it _seems_ to achieve a similar thing?

I'm not sure which of this option is better here.

Reply:
@pnowojski , thanks for your reminder.

It's a good advice. The fail will close stream, it may be slow, so it should hold the lock.

I prefer `enqueue(ChannelStateWriteRequest.abort(...));`, if there is any problem with fail it should not affect the task thread